### PR TITLE
Minor galaxian.cpp changes and fixes

### PR DESCRIPTION
**Pull request by jackson2k2**
The first part of a giant journey on the 0.2xx timeline that will consist of major changes to galaxian.cpp, eventually ending up with a complete migration of scramble.cpp, scobra.cpp, and galaxold.cpp to the new driver altogether. Any kind of feedback is welcome, especially considering how I don't know if how I handled changing the pixel clock scale in galaxian.cpp for Sidam bootlegs is acceptable.

This PR consists of minor commits to correct Take Off, Explorer, Amigo, Jump Bug, Scramble, Jungler on Scramble, The End, War of The Bugs, and Mandinga. Mostly copypasted from the commit logs:

- Correct sync in Take Off, Explorer, and set 2 of Amigo. Sidam bootlegs of Konami games have a 12MHz pixel clock with video being scaled by two pixels instead of 3.
- Correct Jump Bug background hardware behavior, as per real hardware. Stars actually show up until the 232nd pixel.
- Correct bullet behavior in Scramble. Shells are now two pixels long, and have been verified on real hardware.
- Fix(?) clocks on jungsub. The Subelectro 113 board has no XTALs, and I am assuming the CLK lines come from the pixel clock on the Konami L-1200-2 board.
- Various memory map corrections; most of them to add mirrors. These are mostly guesses, and are not precisely right. Feedback appreciated if you have corrections to make.
- War of Bugs and clones have stars disabled, and so does likely Explorer. War of the Bugs has a pin on LS366 '2B' cut off, but not the LS259 that controls it. Therefore, I have made a "null_draw_background" routine to reflect this instead of unmapping the write. It is unknown how stars work on Explorer and Take Off, but Take Off has no code to write to the star generator. So, I am going to imply that none of them have one altogether until I get real hardware verification.
- Given Take Off's hardware hierarchy, correct the year from 1980 to 1981 as a guess.
- Fix protection in The End and Scramble. theend now shows the HIT UFO screen after the CHANCE TIME round.
- Correct machine setup for Take Off, Explorer, and set 2 of Amigo. All are now correctly set up as per the 12 MHz pixel clock.
- Mark Explorer and set 2 of Amigo as MACHINE_IMPERFECT_SOUND. Explorer requires the same sound timer as Take Off, which I have no idea how it works. The latter probably has a similar case, although it might be different from Explorer.
- Mark Take Off as MACHINE_WRONG_COLORS. See the reference video for more; some colors that are purple with the default palette_init are green on real hardware.
- Fix sound in Mandinga by adding sound CPU ROMs from Amidar; mark them as BAD_DUMP. Chances that they may be different are unknown.
- Correct background hardware on SF-X and Monster Zero to now show stars and RGB backgrounds. Colors may not be correct, and the current schematics scans are not easy to read.
- Converted galaxian.cpp to use MCFG_SCREEN_VBLANK_CALLBACK.
- turpins is likely more derived from Super Cobra hardware than Scramble hardware. Change the set description to match this statement.
- Fix(?) video on Scramble on Galaxian bootlegs in galaxold.cpp to use regular video inits; also fix sound in scramb2 and scramb3 in galaxold.cpp. It is unlikely that any of the Scramble on Galaxian bootlegs have any changes related to video, and any writes to the background color generator may simply be leftovers. The top board on Hammy's scramb3 board does not show anything video related that maybe be questionable.

As for the constexpr pollution in the headers: I do acknowledge this, but sound/galaxian.cpp won't like it if I move them to the driver itself.

### CODE REVIEW

**Comment by Osso13** [comment]
Nice!
Re mandinga: according to the person who forwarded the dump to the DU, both PCBs located didn't output sounds and ROMs were the same in both.

File: src/mame/drivers/galaxian.cpp [hunk 1]
```diff
@@ -6864,17 +6926,33 @@ void galaxian_state::init_victoryc()
...
 {
-	init_galaxian();
-
 	address_space &space = m_maincpu->space(AS_PROGRAM);
 
-	/* needs a full 2k of RAM */
+	/* same as galaxian... */
+	init_galaxian();
+
+	/* ..but needs a full 2k of RAM */
```
Comment (cuavas) [inline src/mame/drivers/galaxian.cpp]: This kind of change is undesirable - it's better to minimise the scope of variables and keep the delcaration/initialisation close to the point of use.  Also, the "same as galaxian" comment is pointless - it tells you no more than the line it's attached to.  Comments like this just add clutter.

Same applies to the other init functions.

The other issue is that this is a legacy code pattern that we're trying to discourage.  Things should be converted to use derived address maps, not add more code to adjust mapping in init methods.

File: src/mame/drivers/galaxold.cpp [hunk 2]
```diff
@@ -422,15 +423,15 @@ void galaxold_state::scrambler_map(address_map &map)
 	map(0x6805, 0x6805).w("cust", FUNC(galaxian_sound_device::fire_enable_w));
 	map(0x6806, 0x6807).w("cust", FUNC(galaxian_sound_device::vol_w));
 	map(0x7000, 0x7000).portr("IN2").w(FUNC(galaxold_state::galaxold_nmi_enable_w));
-//  AM_RANGE(0x7001, 0x7001)
+// AM_RANGE(0x7001, 0x7001)
 	map(0x7002, 0x7002).w(FUNC(galaxold_state::galaxold_coin_counter_w));
-	map(0x7003, 0x7003).w(FUNC(galaxold_state::scrambold_background_enable_w));
+	map(0x7003, 0x7003).nopw // background write, may not exist on hardware
```
Comment (cuavas) [inline src/mame/drivers/galaxold.cpp]: Um, how is this line supposed to compile?

File: src/mame/drivers/galaxold.cpp [hunk 3]
```diff
@@ -341,7 +341,7 @@ void galaxold_state::scramblb_map(address_map &map)
 	map(0x7000, 0x7000).portr("IN2");
 	map(0x7001, 0x7001).w(FUNC(galaxold_state::galaxold_nmi_enable_w));
 	map(0x7002, 0x7002).w(FUNC(galaxold_state::galaxold_coin_counter_w));
-	map(0x7003, 0x7003).w(FUNC(galaxold_state::scrambold_background_enable_w));
+	map(0x7003, 0x7003).nopw // background write, may not exist on hardware
```
Comment (cuavas) [inline src/mame/drivers/galaxold.cpp]: This line can't possibly compile.

File: src/mame/drivers/galaxold.cpp [hunk 4]
```diff
@@ -368,14 +368,15 @@ void galaxold_state::scramb_common_map(address_map &map)
 	map(0x5080, 0x50ff).ram();
 	map(0x5800, 0x5fff).r(FUNC(galaxold_state::scramb2_protection_r)); // must return 0x25
 	map(0x6000, 0x6007).r(FUNC(galaxold_state::scramb2_port0_r)); // reads from 8 addresses, 1 bit per address
+	map(0x6004, 0x6007).w("cust", FUNC(galaxian_sound_device::lfo_freq_w));
 	map(0x6800, 0x6807).r(FUNC(galaxold_state::scramb2_port1_r)); // reads from 8 addresses, 1 bit per address
 	map(0x6801, 0x6801).w(FUNC(galaxold_state::galaxold_nmi_enable_w));
+	map(0x6803, 0x6803).nopw // background write, may not exist on hardware
```
Comment (cuavas) [inline src/mame/drivers/galaxold.cpp]: This can't compile.

File: src/mame/drivers/galaxold.cpp [hunk 5]
```diff
@@ -2383,13 +2370,6 @@ MACHINE_CONFIG_START(galaxold_state::scrambler)
 	/* basic machine hardware */
 	MCFG_DEVICE_MODIFY("maincpu")
 	MCFG_DEVICE_PROGRAM_MAP(scrambler_map)
-
-	/* video hardware */
-	MCFG_PALETTE_MODIFY("palette")
-	MCFG_PALETTE_ENTRIES(32+2+64+1)  /* 32 for the characters, 2 for the bullets, 64 for the stars, 1 for background */
-
-	MCFG_PALETTE_INIT_OWNER(galaxold_state,scrambold)
-	MCFG_VIDEO_START_OVERRIDE(galaxold_state,scrambold)
```
Comment (cuavas) [inline src/mame/drivers/galaxold.cpp]: What's the actual reason for removing these?  It's not clear from the description.

File: src/mame/drivers/galaxian.cpp [hunk 6]
```diff
@@ -889,52 +887,73 @@ WRITE8_MEMBER(galaxian_state::theend_coin_counter_w)
...
-		case 0xf09:     m_protection_result = 0xff; break;
-		case 0xa49:     m_protection_result = 0xbf; break;
-		case 0x319:     m_protection_result = 0x4f; break;
-		case 0x5c9:     m_protection_result = 0x6f; break;
-
-		/* scrambls */
-		case 0x246:     m_protection_result ^= 0x80;    break;
-		case 0xb5f:     m_protection_result = 0x6f; break;
+		/* TODO: theend writes several, other mystery sequences */
+		case 0x6:
```
Comment (cuavas) [inline src/mame/drivers/galaxian.cpp]: This kind of indentation is bad.  There's no need for braces when there are not initialisations or things that need to be unwound, and the case labels don't create scope blocks.  To be honest, it was a lot easier to read in tabular form, but if you must break it up lines, do it in a way that's less visually cluttered, e.g.
```
    switch (op)
    {
    case 0x6:
        m_protection_result ^= 0x80; // scrambles
        break;
    case 0x9:
        m_protection_result ^= 0x80; // scrambles
    ...
    }
```

**Comment by MooglyGuy** [comment]
It should be merged, but why is it failing Travis?

**Comment by Tafoid** [comment]
Someone would need to pull the changes and verify it compiles cleanly locally... Travis appears to be and error on some CLANG related issue with other code at the time.  Since these checks only check core files, it is impossible to know if the code compiles cleanly.   Even then,  I would humbly suggest waiting until release freeze where I will have more time to test all the migrations and changes.  While you could argue it should have gone in already and been worked on but the OP has had nearly a month to respond/make any effort to address concerns.

**Comment by MooglyGuy** [comment]
Good call. It can be merged in for the 0.202 cycle.

**Comment by cuavas** [comment]
The trouble is it changes a bunch of things without explanation, and goes against the direction we're trying to move things by doing things in init functions that should be in derived address maps.  It  really needs to be split up and explained properly.

**Comment by ajrhacker** [comment]
I've started repackaging portions of this commit with a few improvements of my own.

**Pull request closed without merging**